### PR TITLE
Substitution_formatters: Add support for %DOWNSTREAM_{LOCAL,PEER}_{DNS,IP}_SAN%

### PR DESCRIPTION
**Pull request by vikaschoudhary16**
Commit Message: Adding new commands:
- %DOWNSTREAM_LOCAL_DNS_SANS%
- %DOWNSTREAM_LOCAL_IP_SANS%
- %DOWNSTREAM_PEER_DNS_SANS%
- %DOWNSTREAM_PEER_IP_SANS%

Addressing https://github.com/envoyproxy/envoy/issues/14570

Additional Description:
Risk Level:
Testing:
Docs Changes:
Release Notes:
Platform Specific Features:
[Optional Runtime guard:]
[Optional Fixes #Issue]
[Optional Fixes commit #PR or SHA]
[Optional Deprecated:]
[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):]

### CODE REVIEW

**Comment by kyessenov** [comment]
/assign-from @envoyproxy/envoy-maintainers 

This PR needs a doc change to document these new formatter I think.

File: envoy/ssl/connection.h [hunk 1]
```diff
@@ -103,6 +103,18 @@ class ConnectionInfo {
    **/
   virtual absl::Span<const std::string> dnsSansLocalCertificate() const PURE;
 
+  /**
+   * @return absl::Span<const std::string> the IP entries in the SAN field of the peer certificate.
+   *         Returns {} if there is no peer certificate, or no SAN field, or no IPs.
+   **/
+  virtual absl::Span<const std::string> ipSansPeerCertificate() const PURE;
```
Comment (kyessenov) [inline envoy/ssl/connection.h:110]: Pass-by comment: is it worth presenting them as IPs not strings? Do you need to do some wildcard matching?

Comment (ggreenway) [reply envoy/ssl/connection.h:110]: I'd say keep them as strings, unless there's a good reason to convert them. They start out as strings in the certificate, and right now we don't have to parse them at all.

On the other hand, there may be subtle formatting differences that could be avoided by converting them to IP addresses internally.

**Review by ggreenway — CHANGES_REQUESTED** [review]
Needs docs and a release note

/wait

File: source/common/formatter/substitution_formatter.cc [hunk 2]
```diff
@@ -1424,6 +1424,24 @@ const StreamInfoFormatter::FieldExtractorLookupTbl& StreamInfoFormatter::getKnow
                                                          ",");
                                   });
                             }}},
+                          {"DOWNSTREAM_PEER_DNS_SANS",
```
Comment (ggreenway) [inline source/common/formatter/substitution_formatter.cc:1427]: The other SAN formatters use `_SAN` not `_SANS` (plural). The plural is more correct, but maybe we should keep them consistent?

File: test/extensions/transport_sockets/tls/ssl_socket_test.cc [hunk 3]
```diff
@@ -97,6 +100,10 @@ class TestUtilOptionsBase {
     expected_client_cert_uri_ = {expected_client_cert_uri};
   }
 
+  void setExpectedClientIpSans(const std::vector<std::string>& expected_client_ip_sans) {
```
Comment (ggreenway) [inline test/extensions/transport_sockets/tls/ssl_socket_test.cc:103]: I don't see this used. Can you add some test coverage using this?

**Review by ggreenway — CHANGES_REQUESTED** [review]
This is looking good overall.

Please document the new formatters here: https://github.com/envoyproxy/envoy/blob/main/docs/root/configuration/observability/access_log/usage.rst

/wait

File: test/extensions/transport_sockets/tls/ssl_socket_test.cc (hunk 3 above)
Comment (vikaschoudhary16) [reply test/extensions/transport_sockets/tls/ssl_socket_test.cc:103]: done

File: source/common/formatter/substitution_formatter.cc (hunk 2 above)
Comment (vikaschoudhary16) [reply source/common/formatter/substitution_formatter.cc]: done

**Review by ggreenway — APPROVED** [review]
LGTM

**Pull request merged**